### PR TITLE
[SPARK-17410] [SPARK-17284] Move Hive-generated Stats Info to HiveClientImpl

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

After we adding a new field `stats` into `CatalogTable`, we should not expose Hive-specific Stats metadata to `MetastoreRelation`. It complicates all the related codes. It also introduces a bug in `SHOW CREATE TABLE`. The statistics-related table properties should be skipped by `SHOW CREATE TABLE`, since it could be incorrect in the newly created table. See the Hive JIRA: https://issues.apache.org/jira/browse/HIVE-13792

Also fix the issue to fill Hive-generated RowCounts to our stats. 

This PR is to handle Hive-specific Stats metadata in `HiveClientImpl`. 
### How was this patch tested?

Added a few test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
Still need a test case for verifying alter table drop/add partitions

**Comment by gatorsmile** [comment]
Found bugs in the master and 2.0 branch when adding alter table drop/add partitions.

**Comment by gatorsmile** [comment]
... Very surprised about Hive... Any `ALTER TABLE SET/UNSET TBLPROPERTIES` statements can invalidate the Hive-generated statistics...

```Scala
hiveClient.runSqlHive(s"ANALYZE TABLE $oldName COMPUTE STATISTICS")
hiveClient.runSqlHive(s"DESCRIBE FORMATTED $oldName").foreach(println)
```
```
Table Parameters:	 	 
	COLUMN_STATS_ACCURATE	true                
	numFiles            	1                   
	numRows             	500                 
	rawDataSize         	5312                
	spark.sql.statistics.numRows	500                 
	spark.sql.statistics.totalSize	5812                
	totalSize           	5812                
	transient_lastDdlTime	1473610039          
```
```Scala
hiveClient.runSqlHive(s"ALTER TABLE $oldName SET TBLPROPERTIES ('foofoo' = 'a')")
hiveClient.runSqlHive(s"DESCRIBE FORMATTED $oldName").foreach(println)
```
```
Table Parameters:	 	 
	COLUMN_STATS_ACCURATE	false               
	foofoo              	a                   
	last_modified_by    	xiaoli              
	last_modified_time  	1473610039          
	numFiles            	1                   
	numRows             	-1                  
	rawDataSize         	-1                  
	spark.sql.statistics.numRows	500                 
	spark.sql.statistics.totalSize	5812                
	totalSize           	5812                
	transient_lastDdlTime	1473610039          
```

**Comment by gatorsmile** [comment]
cc @hvanhovell @cloud-fan Now, the code is ready for review.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 1]
```diff
@@ -259,6 +259,156 @@ class StatisticsSuite extends QueryTest with TestHiveSingleton with SQLTestUtils
...
+
+      // ALTER TABLE UNSET TBLPROPERTIES invalidates some contents of Hive specific statistics
+      // This is triggered by the Hive alterTable API
+      val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+      val describeResult = hiveClient.runSqlHive(s"DESCRIBE FORMATTED $tabName")
+      assert(StringUtils.filterPattern(describeResult, "*numRows\\s+-1*").nonEmpty)
+    }
+  }
+
+  test("add/drop partitions - managed table") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: FYI, when we drop partitions of EXTERNAL tables, `ANALYZE TABLE` is unable to exclude them from statistics. This should be fixed with https://issues.apache.org/jira/browse/SPARK-17129, if my understanding is right.

**Comment by gatorsmile** [comment]
@hvanhovell @cloud-fan Could you help me review this PR? https://github.com/apache/spark/pull/15090 is changing the same code path for column-level statistics.

Thanks!

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
Will this change break existing behaviour? The `MetastoreRelation` can have table statistics if the hive table is already analyzed.

BTW, I'd like to have this behaviour:

1. `HiveClient.getTable` should return a `CatalogTable` with stats if this table has stats properties of hive.
2. `HiveExternalCatalog.getTable` should get the table via hive client, and overwrite the stats if this table has stats properties of spark, i.e. we trust spark rather than hive.

Any ideas?

**Comment by gatorsmile** [comment]
It does not break the existing behavior. If the MetastoreRelation has the Hive-generated table statistics, we create a statistics [here](https://github.com/gatorsmile/spark/blob/9e18ba104527d2bb14331f4b51194002dabb2556/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L405-L422). If we have Spark-generated statistics, we overwrite the hive-generated one in [restoreTableMetadata](https://github.com/apache/spark/blob/d6eede9a36766e2d2294951b054d7557008a5662/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L482-L484). 

Thus, the current code completely matches what you wants. : )

**Comment by gatorsmile** [comment]
Let me write a test case to ensure this correctly works and also put more comments in the code.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala [hunk 2]
```diff
@@ -111,23 +111,8 @@ private[hive] case class MetastoreRelation(
   @transient override lazy val statistics: Statistics = {
     catalogTable.stats.getOrElse(Statistics(
       sizeInBytes = {
-        val totalSize = hiveQlTable.getParameters.get(StatsSetupConst.TOTAL_SIZE)
-        val rawDataSize = hiveQlTable.getParameters.get(StatsSetupConst.RAW_DATA_SIZE)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: In the master branch, we do not use Hive-generated `numRows`... Let me fix it in this PR.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -378,6 +380,47 @@ private[hive] class HiveClientImpl(
...
 
+      val excludedTableProperties = Set(
+        // The following are hive-generated statistics fields. Currently, only total_size and
+        // row_count are used to populate the dedicated field `stats`.
+        // TODO: stats should include all the other three fields.
+        StatsSetupConst.COLUMN_STATS_ACCURATE,
+        StatsSetupConst.NUM_FILES,
+        StatsSetupConst.NUM_PARTITIONS,
+        StatsSetupConst.ROW_COUNT,
+        StatsSetupConst.RAW_DATA_SIZE,
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: how about we only handle the 2 we need? i.e. `TOTAL_SIZE` and `RAW_DATA_SIZE`. Then we don't need to do an extra `getTable` call in `alterTable`, which may cause performance regression.

Ideally the rule is, we only drop the hive properties that we moved to other places, so that we can reconstruct them without an extra `getTable` call.

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: That means, we will overwrite the Hive-generated statistics `TOTAL_SIZE` and `RAW_DATA_SIZE` by our statistics. This could be a surprise to the users who are using both Hive and Spark on the same data sets, when they issues an alter table from Spark.

If we want to do not hide the other Hive-specific fields (e.g., `NUM_FILES`, `NUM_PARTITIONS`), `SHOW CREATE TABLE` needs to explicitly exlude them, like what we did in the PR: https://github.com/apache/spark/pull/14855. Does this sound OK?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -378,6 +380,51 @@ private[hive] class HiveClientImpl(
...
+      )
+
+      val filteredProperties = properties.filterNot {
+        case (key, _) => excludedTableProperties.contains(key)
+      }
+      val comment = properties.get("comment")
+
+      val totalSize = properties.get(StatsSetupConst.TOTAL_SIZE).map(s => BigInt(s.toLong))
+      val rawDataSize = properties.get(StatsSetupConst.RAW_DATA_SIZE).map(s => BigInt(s.toLong))
+      val rowCount = properties.get(StatsSetupConst.ROW_COUNT).map(s => BigInt(s.toLong)) match {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Here, we also utilize Hive-generated row counts when users have not run ANALYZE TABLE through Spark.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: ah I see. So our targets are:

1. recognize hive statistics, i.e. we should set the `CatalogTable.stats` according to hive stats properties
2. don't overwrite hive stats properties.
3. SHOW CREATE TABLE shouldn't print hive stats properties.

My proposal: In `HiveClientImpl`, set `CatalogTable.stats` by hive stats properties, and still keep them in table properties. In SHOW CREATE TABLE, hide the hive stats properties.

Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: `CatalogTable` have a field `unsupportedFeatures`, can we extend it to hide this kind of hive specific properties which are only useful in alter table?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Using `unsupportedFeatures` sounds a pretty good idea! Let me make a try. Thanks!

**Comment by gatorsmile** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 5]
```diff
@@ -259,6 +260,230 @@ class StatisticsSuite extends QueryTest with TestHiveSingleton with SQLTestUtils
...
+        sql(
+          s"""
+             |INSERT OVERWRITE TABLE $managedTable
+             |partition (ds='$ds',hr='$hr')
+             |SELECT 1, 'a'
+           """.stripMargin)
+      }
+
+      checkStats(
+        managedTable, isDataSourceTable = false, hasSizeInBytes = false, expectedRowCounts = None)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Actually, this is a surprise to me. We did not use Hive-generated statistics. I found the table-level statistics is missing. We need to get the statistics info from the properties for each partition and then add them up. Will submit a separate PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 6]
```diff
@@ -134,7 +134,7 @@ case class CatalogTable(
     viewOriginalText: Option[String] = None,
     viewText: Option[String] = None,
     comment: Option[String] = None,
-    unsupportedFeatures: Seq[String] = Seq.empty) {
+    unsupportedFeatures: Map[String, String] = Map.empty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: should we give it a new name? hive stats properties are not unsupported but ignored...

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: or we may just add a new field

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: 👍 Let me add a new field called `ignoredProperties`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 7]
```diff
@@ -216,16 +218,15 @@ class StatisticsSuite extends QueryTest with TestHiveSingleton with SQLTestUtils
...
         expectedRowCounts = None)
       sql(s"INSERT INTO TABLE $textTable SELECT * FROM src")
-      checkStats(
-        textTable,
-        isDataSourceTable = false,
-        hasSizeInBytes = false,
-        expectedRowCounts = None)
+
+      val fetchedStats = checkStats(
+        textTable, isDataSourceTable = false, hasSizeInBytes = true, expectedRowCounts = None)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: we should also update the comment in L212

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 8]
```diff
@@ -259,6 +260,230 @@ class StatisticsSuite extends QueryTest with TestHiveSingleton with SQLTestUtils
...
+      val pattern = new Regex(s"""$propKey\\s+(-?\\d+)""")
+      val pattern(value) = str.head.trim
+      Option(BigInt(value))
+    }
+  }
+
+  test("get statistics when not analyzed in both Hive and Spark") {
+    val tabName = "tab1"
+    val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+    hiveClient.runSqlHive("set hive.stats.autogather=false")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: we should reset this conf back after test

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 9]
```diff
@@ -259,6 +260,230 @@ class StatisticsSuite extends QueryTest with TestHiveSingleton with SQLTestUtils
...
+    // statistics info is not contained in the metadata of the original table
+    assert(Seq(StatsSetupConst.COLUMN_STATS_ACCURATE,
+      StatsSetupConst.NUM_FILES,
+      StatsSetupConst.NUM_PARTITIONS,
+      StatsSetupConst.ROW_COUNT,
+      StatsSetupConst.RAW_DATA_SIZE,
+      StatsSetupConst.TOTAL_SIZE).forall(!tableMetadata.contains(_)))
+
+    if (analyzedByHive) {
+      assert(StringUtils.filterPattern(describeResult1, "*numRows\\s+500*").nonEmpty)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: why we test the stats at hive side not spark side?

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Hive does not recognize Spark-generated statistics. Thus, if `analyzedByHive` is `false`, we do not expect to see the collected `numRows`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: This is not related to the code. Let me remove it. 

No matter whether this is true or false, totalSize is a positive value. Also tried `INSERT OVERWRITE`. Still the same. Maybe a bug in Hive?

Originally, I want to see the impact. It does not behave consistently with my installed Hive version.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Done.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 10]
```diff
@@ -378,6 +380,50 @@ private[hive] class HiveClientImpl(
...
+      val stats =
+        // When table is external, `totalSize` is always zero, which will influence join strategy
+        // so when `totalSize` is zero, use `rawDataSize` instead. When `rawDataSize` is also zero,
+        // return None. Later, we will use the other ways to estimate the statistics.
+        if (totalSize.isDefined && totalSize.get > 0L) {
+          Some(Statistics(sizeInBytes = totalSize.get, rowCount = rowCount))
+        } else if (rawDataSize.isDefined && rawDataSize.get > 0) {
+          Some(Statistics(sizeInBytes = rawDataSize.get, rowCount = rowCount))
+        } else {
+          None
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: shall we respect `rowCount` in this case?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 11]
```diff
@@ -378,6 +380,50 @@ private[hive] class HiveClientImpl(
 
       val properties = Option(h.getParameters).map(_.asScala.toMap).orNull
 
+      // Hive-generated Statistics are also recorded in ignoredProperties
+      val ignoredProperties = scala.collection.mutable.Map.empty[String, String]
+      properties.collect {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: how about
```
for (key <- HiveStatisticsProperties; value <- properties.get(key)) {
  ignoredProperties += key -> value
}
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 12]
```diff
@@ -427,7 +473,14 @@ private[hive] class HiveClientImpl(
...
   override def alterTable(tableName: String, table: CatalogTable): Unit = withHiveState {
-    val hiveTable = toHiveTable(table)
+    // getTableOption removes all the Hive-specific properties. Here, we fill them back to ensure
+    // these properties are still available to the others that share the same Hive metastore.
+    // If users explicitly alter these Hive-specific properties through ALTER TABLE DDL, we respect
+    // these user-specified values.
+    val hiveSpecificProps = table.ignoredProperties.filter {
+      case (key, _) => HiveStatisticsProperties.contains(key)
+    }
+    val hiveTable = toHiveTable(table.copy(properties = table.properties ++ hiveSpecificProps))
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: `table.properties ++ hiveSpecificProps`, by doing this, we won't respect user-specified Hive-specific properties, which are in `table.properties`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 13]
```diff
@@ -427,7 +473,14 @@ private[hive] class HiveClientImpl(
   }
 
   override def alterTable(tableName: String, table: CatalogTable): Unit = withHiveState {
-    val hiveTable = toHiveTable(table)
+    // getTableOption removes all the Hive-specific properties. Here, we fill them back to ensure
+    // these properties are still available to the others that share the same Hive metastore.
+    // If users explicitly alter these Hive-specific properties through ALTER TABLE DDL, we respect
+    // these user-specified values.
+    val hiveSpecificProps = table.ignoredProperties.filter {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: do we really need to do `filter` here? I think the semantic of `ignoredProperties` is: ignore these properties in Spark SQL, but get them back when saving the metastore

**Comment by gatorsmile** [comment]
Recently, CBO introduces many changes in this part. Let me revisit it. Thank you for the reviews! @cloud-fan

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 14]
```diff
@@ -414,6 +415,50 @@ private[hive] class HiveClientImpl(
...
+      // NOTE: getting `totalSize` directly from params is kind of hacky, but this should be
+      // relatively cheap if parameters for the table are populated into the metastore.
+      // Currently, only totalSize, rawDataSize, and row_count are used to build the field `stats`
+      // TODO: stats should include all the other two fields (`numFiles` and `numPartitions`).
+      // (see StatsSetupConst in Hive)
+      val stats =
+      // When table is external, `totalSize` is always zero, which will influence join strategy
+      // so when `totalSize` is zero, use `rawDataSize` instead. When `rawDataSize` is also zero,
+      // return None. Later, we will use the other ways to estimate the statistics.
+      if (totalSize.isDefined && totalSize.get > 0L) {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: the indention is wrong

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala [hunk 15]
```diff
@@ -325,26 +325,24 @@ class ShowCreateTableSuite extends QueryTest with SQLTestUtils with TestHiveSing
...
         lastAccessTime = 0L,
-        properties = table.properties.filterKeys(!nondeterministicProps.contains(_))
+        properties = table.properties.filterKeys(!nondeterministicProps.contains(_)),
+        stats = None,
+        ignoredProperties = Map.empty
       )
     }
 
+    val e = normalize(actual)
+    val m = normalize(expected)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ShowCreateTableSuite.scala]: remove this?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 16]
```diff
@@ -175,7 +178,7 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
       sql(s"INSERT INTO TABLE $textTable SELECT * FROM src")
       checkTableStats(
         textTable,
-        hasSizeInBytes = false,
+        hasSizeInBytes = true,
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: why the behavior is changed?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 17]
```diff
@@ -414,6 +415,50 @@ private[hive] class HiveClientImpl(
...
+      )
+
+      val filteredProperties = properties.filterNot {
+        case (key, _) => excludedTableProperties.contains(key)
+      }
+      val comment = properties.get("comment")
+
+      val totalSize = properties.get(StatsSetupConst.TOTAL_SIZE).map(BigInt(_))
+      val rawDataSize = properties.get(StatsSetupConst.RAW_DATA_SIZE).map(BigInt(_))
+      def rowCount = properties.get(StatsSetupConst.ROW_COUNT).map(BigInt(_)) match {
```
Comment (wzhfy) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: why use def?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 16 above)
Comment (wzhfy) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Because now in `HiveClientImpl.getTableOption`, we respect Hive's stats.

Comment (wzhfy) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Hive will alter totalSize after inserting data.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveComparisonTest.scala [hunk 18]
```diff
@@ -192,13 +192,7 @@ abstract class HiveComparisonTest
     "last_modified_by",
     "last_modified_time",
     "Owner:",
-    "COLUMN_STATS_ACCURATE",
     // The following are hive specific schema parameters which we do not need to match exactly.
-    "numFiles",
-    "numRows",
-    "rawDataSize",
-    "totalSize",
-    "totalNumberFiles",
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveComparisonTest.scala]: `TextMetaDataFormatter` and `JsonMetaDataFormatter` insert these info based on `numFiles`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 19]
```diff
@@ -232,7 +446,8 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
         checkTableStats(parquetTable, hasSizeInBytes = true, expectedRowCounts = Some(500))
       }
       withSQLConf("spark.sql.hive.convertMetastoreOrc" -> "true") {
-        checkTableStats(orcTable, hasSizeInBytes = false, expectedRowCounts = None)
+        // We still can get tableSize from Hive before Analyze
+        checkTableStats(orcTable, hasSizeInBytes = true, expectedRowCounts = None)
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: A good question. This is from Hive. : ( I did not investigate the root cause inside Hive.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveComparisonTest.scala (hunk 18 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveComparisonTest.scala]: I think we should keep it unchanged.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 17 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: only used once. We also can use lazy val

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 20]
```diff
@@ -215,6 +218,217 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+  }
+
+  test("alter table UNSET TBLPROPERTIES after analyze table") {
+    Seq(true, false).foreach { analyzedBySpark =>
+      val tabName = "tab1"
+      withTable(tabName) {
+        createNonPartitionedTable(tabName, analyzedByHive = true, analyzedBySpark = analyzedBySpark)
+        val fetchedStats1 = checkTableStats(
+          tabName, hasSizeInBytes = true, expectedRowCounts = Some(500))
+        sql(s"ALTER TABLE $tabName UNSET TBLPROPERTIES ('prop1')")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: The prop values are not changed after set/unset in Hive 2.x

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 16 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: It sounds like Hive does online stats updates.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 21]
```diff
@@ -414,6 +415,50 @@ private[hive] class HiveClientImpl(
...
+      )
+
+      val filteredProperties = properties.filterNot {
+        case (key, _) => excludedTableProperties.contains(key)
+      }
+      val comment = properties.get("comment")
+
+      val totalSize = properties.get(StatsSetupConst.TOTAL_SIZE).map(BigInt(_))
+      val rawDataSize = properties.get(StatsSetupConst.RAW_DATA_SIZE).map(BigInt(_))
+      lazy val rowCount = properties.get(StatsSetupConst.ROW_COUNT).map(BigInt(_)) match {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: 1. I think we can just use val, no need to bother about performance here.
2. can be simplified to `xxx.filter(_ >= 0)`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 22]
```diff
@@ -215,6 +218,215 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+      withTable(tabName) {
+        createNonPartitionedTable(tabName, analyzedByHive = true, analyzedBySpark = analyzedBySpark)
+        val fetchedStats1 = checkTableStats(
+          tabName, hasSizeInBytes = true, expectedRowCounts = Some(500))
+        sql(s"ALTER TABLE $tabName SET TBLPROPERTIES ('foo' = 'a')")
+        val fetchedStats2 = checkTableStats(
+          tabName, hasSizeInBytes = true, expectedRowCounts = Some(500))
+        assert(fetchedStats1 == fetchedStats2)
+
+        val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: this appeared many times, we can create a method

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
seems like a valid test failure

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merging to master.
